### ISSUE TITLE
[All Platforms] Automatic widget choice based on Platform

### ISSUE THREAD

**Issue by YazeedAlKhalaf** [flutter/flutter#62260]
What I suggest is that Flutter adds new widgets type, other than Cupertino and Material.

Maybe call it an adaptive or similar name. The purpose will be in writing only once and the Platform checks will be done in the original code not by the dev. It will make it easier. Imagine having to make each screen twice, but if the adaptive library is there. It will be a piece of cake, half of the work is gone.

It is not something that is hard to do but will save developers a lot of time.

The concept will be explained here:

<details>
<summary>Code Examples</summary>

Current Way:
```dart
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:instagram_clone/src/ui/global/app_colors.dart';
import 'package:stacked/stacked.dart';

import './home_view_model.dart';

class HomeView extends StatefulWidget {
  @override
  _HomeViewState createState() => _HomeViewState();
}

class _HomeViewState extends State<HomeView> {
  @override
  Widget build(BuildContext context) {
    return Platform.isIOS ? HomeViewCupertino() : HomeViewMaterial();
  }
}

class HomeViewMaterial extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return ViewModelBuilder<HomeViewModel>.reactive(
      viewModelBuilder: () => HomeViewModel(),
      builder: (
        BuildContext context,
        HomeViewModel model,
        Widget child,
      ) {
        return Scaffold(
          body: model.chooseScreen(),
          floatingActionButtonLocation:
              FloatingActionButtonLocation.centerDocked,
          floatingActionButton: FloatingActionButton(
            onPressed: () async => await model.navigateToCreatePostView(),
            tooltip: 'Create Post',
            child: Icon(Icons.add_a_photo),
            elevation: 2.0,
          ),
          bottomNavigationBar: BottomAppBar(
            child: Row(
              mainAxisSize: MainAxisSize.max,
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: [
                IconButton(
                  iconSize: 30.0,
                  color: model.currentTab == 0 ? primaryColor : lynchColor,
                  icon: Icon(Icons.home),
                  onPressed: () {
                    model.setCurrentTab(0);
                  },
                ),
                IconButton(
                  iconSize: 30.0,
                  color: model.currentTab == 1 ? primaryColor : lynchColor,
                  icon: Icon(Icons.search),
                  onPressed: () {
                    model.setCurrentTab(1);
                  },
                ),
                IconButton(
                  iconSize: 30.0,
                  color: model.currentTab == 2 ? primaryColor : lynchColor,
                  icon: Icon(Icons.notifications),
                  onPressed: () {
                    model.setCurrentTab(2);
                  },
                ),
                IconButton(
                  iconSize: 30.0,
                  color: model.currentTab == 3 ? primaryColor : lynchColor,
                  icon: Icon(Icons.account_circle),
                  onPressed: () {
                    model.setCurrentTab(3);
                  },
                ),
              ],
            ),
            shape: CircularNotchedRectangle(),
            color: Colors.white,
          ),
        );
      },
    );
  }
}

// Consider this cupertino widgets :)
class HomeViewCupertino extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return ViewModelBuilder<HomeViewModel>.reactive(
      viewModelBuilder: () => HomeViewModel(),
      builder: (
        BuildContext context,
        HomeViewModel model,
        Widget child,
      ) {
        return Scaffold(
          body: model.chooseScreen(),
          floatingActionButtonLocation:
              FloatingActionButtonLocation.centerDocked,
          floatingActionButton: FloatingActionButton(
            onPressed: () async => await model.navigateToCreatePostView(),
            tooltip: 'Create Post',
            child: Icon(Icons.add_a_photo),
            elevation: 2.0,
          ),
          bottomNavigationBar: BottomAppBar(
            child: Row(
              mainAxisSize: MainAxisSize.max,
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: [
                IconButton(
                  iconSize: 30.0,
                  color: model.currentTab == 0 ? primaryColor : lynchColor,
                  icon: Icon(Icons.home),
                  onPressed: () {
                    model.setCurrentTab(0);
                  },
                ),
                IconButton(
                  iconSize: 30.0,
                  color: model.currentTab == 1 ? primaryColor : lynchColor,
                  icon: Icon(Icons.search),
                  onPressed: () {
                    model.setCurrentTab(1);
                  },
                ),
                IconButton(
                  iconSize: 30.0,
                  color: model.currentTab == 2 ? primaryColor : lynchColor,
                  icon: Icon(Icons.notifications),
                  onPressed: () {
                    model.setCurrentTab(2);
                  },
                ),
                IconButton(
                  iconSize: 30.0,
                  color: model.currentTab == 3 ? primaryColor : lynchColor,
                  icon: Icon(Icons.account_circle),
                  onPressed: () {
                    model.setCurrentTab(3);
                  },
                ),
              ],
            ),
            shape: CircularNotchedRectangle(),
            color: Colors.white,
          ),
        );
      },
    );
  }
}
```

New Way:

```dart
import 'package:flutter/adaptive.dart'; // Notice the change here :)
import 'package:instagram_clone/src/ui/global/app_colors.dart';
import 'package:stacked/stacked.dart';

import './home_view_model.dart';

class HomeView extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return ViewModelBuilder<HomeViewModel>.reactive(
      viewModelBuilder: () => HomeViewModel(),
      builder: (
        BuildContext context,
        HomeViewModel model,
        Widget child,
      ) {
        return AdaptiveScaffold(
          body: model.chooseScreen(),
          floatingActionButtonLocation:
              AdaptiveFloatingActionButtonLocation.centerDocked,
          floatingActionButton: AdaptiveFloatingActionButton(
            onPressed: () async => await model.navigateToCreatePostView(),
            tooltip: 'Create Post',
            child: AdaptiveIcon(Icons.add_a_photo),
            elevation: 2.0,
          ),
          bottomNavigationBar: AdaptiveBottomAppBar(
            child: Row(
              mainAxisSize: MainAxisSize.max,
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: [
                AdaptiveIconButton(
                  iconSize: 30.0,
                  color: model.currentTab == 0 ? primaryColor : lynchColor,
                  icon: AdaptiveIcon(Icons.home),
                  onPressed: () {
                    model.setCurrentTab(0);
                  },
                ),
                AdaptiveIconButton(
                  iconSize: 30.0,
                  color: model.currentTab == 1 ? primaryColor : lynchColor,
                  icon: AdaptiveIcon(Icons.search),
                  onPressed: () {
                    model.setCurrentTab(1);
                  },
                ),
                AdaptiveIconButton(
                  iconSize: 30.0,
                  color: model.currentTab == 2 ? primaryColor : lynchColor,
                  icon: AdaptiveIcon(Icons.notifications),
                  onPressed: () {
                    model.setCurrentTab(2);
                  },
                ),
                AdaptiveIconButton(
                  iconSize: 30.0,
                  color: model.currentTab == 3 ? primaryColor : lynchColor,
                  icon: AdaptiveIcon(Icons.account_circle),
                  onPressed: () {
                    model.setCurrentTab(3);
                  },
                ),
              ],
            ),
            shape: AdaptiveCircularNotchedRectangle(),
            color: Colors.white,
          ),
        );
      },
    );
  }
}
```

</details>

I think it will be a lot easier and cleaner if the adaptive widgets are implemented.

Adaptive is just a name suggestion, could be called anything else. But I think the Adaptive keyword is great!

**Comment by EducatedDeer** [comment]
you can try this
https://pub.dev/packages/flutter_platform_widgets

**Comment by YazeedAlKhalaf** [comment]
Tried it. It is good and the owner of the package did a great work. But what I mean is that in Flutter, I feel it is better that this becomes part of the framework. As I mentioned above in Code part:

```dart
import ‘package:flutter/adaptive.dart’;
```

So yea that’s my purpose of this issue.


To track the progress of making this a part of the framework. I could contribute 🤗.

This might even make Flutter greater and more popular.

**Comment by pedromassangocode** [comment]
Related to https://github.com/flutter/flutter/issues/33830

**Comment by goderbauer** [comment]
Duplicate of https://github.com/flutter/flutter/issues/33830.
